Feature 2429 tc pairs members

sethlinden opened the pull request:
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>

Yes, I added a new configuration field for the consensus[] dictionary. Added write_members field. If this is set to FALSE, output for the consensus members will no longer show up in the output file.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Ran a tc_pairs command line passed on the CONSENSUS unit test:

./tc_pairs -adeck /d1/projects/MET/MET_test_data/unit_test/tc_data/adeck/aal132020.dat -bdeck /d1/projects/MET/MET_test_data/unit_test/tc_data/bdeck/bal132020.dat -config ./TCPairsConfig_CONSENSUS -out al132020_CONSENSUS -v 2

After updating ../MET/internal/test_unit/config/TCPairsConfig_CONSENSUS

I also ran the unit test:

unit_tc_pairs.xml

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Run the unit test: 

unit_tc_pairs.xml

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

Yes, I updated the documentation for tc_pairs:

MET/docs/Users_Guide/

tc-pairs.rst

- [x] Do these changes include sufficient testing updates? **[No]**

No new unit tests were created. Just an updated to the config file: TCPairsConfig_CONSENSUS

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

Commit:
Per issue #2429, added conf_key_write_members, write_members to tc-pairs strings. SL

Commit:
Per issue #2429: added WriteMembers to ConsensusInfo and SkipConsensusMembers to TCPairsConfInfo class. SL

Commit:
Per issue #2429: added code to process Consensus[i].WriteMembers and then check those boolean values to add to SkipConsensusMembers. SL. ci-skip-all

Commit:
Per issue #2429: added config parsing for write_members, created new string array SkipConsensusMembers to keep track of members to skip output for. Modified filter_tracks to skip the tracks (consensus members) that are listed in SkipConsensusMembers. SL

Commit:
Per issue #2429, added write_members = TRUE to all consensus models. SL ci-skip-all

Commit:
Merge branch 'develop' into feature_2429_tc_pairs_members

Commit:
Per issue #2429, added modification note to top of file. SL

Commit:
Per issue #2429, in the consensus section added documentation for new field: write_members. SL

JohnHalleyGotway committed:
Update tc_pairs.cc

Making the order of the log messages match the order in which the filtering logic is applied. Putting "Rejected for skip members" first since that's checked first.

JohnHalleyGotway committed:
Update tc_pairs.cc

JohnHalleyGotway committed:
Update tc_pairs_conf_info.cc

Very minor change. Just calling adding all model names in one call rather than looping over them element by element.

JohnHalleyGotway committed:
Update tc-pairs.rst

Fixing mis-spellings and tweaking wording.

JohnHalleyGotway commented on internal/test_unit/config/TCPairsConfig_CONSENSUS:
```diff
@@ -107,42 +107,49 @@ consensus = [
                    "UE35" ];
       required = [];
       min_req  = 36;
+      write_members = TRUE;
```
```suggestion
      write_members = FALSE;
```
Recommend that we set this flag to FALSE to demonstrate in the unit test that this functionality actually works. This is the exactly the sort of case that prompted this issue in the first place. Would be nice to suppress output for all those UE## member tracks.

sethlinden committed:
Update internal/test_unit/config/TCPairsConfig_CONSENSUS

Co-authored-by: John Halley Gotway <johnhg@ucar.edu>

JohnHalleyGotway reviewed (approved):
I approve of these changes.

Note that **consensus.write_members** is NOW a required entry. Adding this new requirement may cause a headache for existing users that already have consensus defined in existing TC-Pairs config files. The good news is that this produces an obvious runtime error:
```
ERROR  : 
ERROR  : Dictionary::lookup_bool() -> lookup failed for name "write_members"
ERROR  : 
```

I had considered refining the Dictionary::lookup_bool() logic to print a warning message about it being missing instead, but worry about modifying the dictionary lookup logic.

JohnHalleyGotway committed:
Per #2429, as decided during the MET development meeting on 2/23/2023, making the specification of consensus.write_members optional. If present in the config file, it'll be used. If not, this warning message is printed:

WARNING:
WARNING: TCPairsConfInfo::process_config() -> "consensus.write_members" is missing. Using default value of true.
WARNING:

We chose to do this because it's going into a minor release (v11.1.0).
We do no want to cause v11.0.0 config files to cause an error when used
in v11.1.0.

JohnHalleyGotway commented:
Proceeding with the merge of this PR. Based on feedback from @j-opatz today, I updated the logic to print a warning message if write_members is missing:
```
WARNING: 
WARNING: TCPairsConfInfo::process_config() -> "consensus.write_members" is missing. Using default value of true.
WARNING: 
```
But I confirmed that warning message goes away once you add write_members as a true/false boolean.
